src/Storages/MergeTree/MergeTreeDataPartWriterWide.cpp

```diff
@@ -101,35 +103,95 @@ void MergeTreeDataPartWriterWide::write(const Block & block,
     for (size_t i = 0; i < columns_list.size(); ++i, ++it)
     {
+        WrittenOffsetColumns & offset_columns = writing_thread_pool ? offset_columns_per_column[i] : offset_columns_per_column.back();
+        const ColumnWithTypeAndName & column = block.getByName(it->name);
+
+        prepareWriteColumn(column.name, *column.type, offset_columns);
+    }
+
+    it = columns_list.begin();
+    for (size_t i = 0; i < columns_list.size(); ++i, ++it)
+    {
+        WrittenOffsetColumns & offset_columns = writing_thread_pool ? offset_columns_per_column[i] : offset_columns_per_column.back();
         const ColumnWithTypeAndName & column = block.getByName(it->name);
 
-        if (permutation)
+        auto write_column_job = [&, i, it]
         {
-            if (primary_key_block.has(it->name))
+            ColumnWriteResult result;
+            if (permutation)
             {
-                const auto & primary_column = *primary_key_block.getByName(it->name).column;
-                writeColumn(column.name, *column.type, primary_column, offset_columns);
-            }
-            else if (skip_indexes_block.has(it->name))
-            {
-                const auto & index_column = *skip_indexes_block.getByName(it->name).column;
-                writeColumn(column.name, *column.type, index_column, offset_columns);
+                if (primary_key_block.has(it->name))
+                {
+                    const auto & primary_column = *primary_key_block.getByName(it->name).column;
+                    result = writeColumn(column.name, *column.type, primary_column, offset_columns);
+                }
+                else if (skip_indexes_block.has(it->name))
+                {
+                    const auto & index_column = *skip_indexes_block.getByName(it->name).column;
+                    result = writeColumn(column.name, *column.type, index_column, offset_columns);
+                }
+                else
+                {
+                    /// We rearrange the columns that are not included in the primary key here; Then the result is released - to save RAM.
+                    ColumnPtr permuted_column = column.column->permute(*permutation, 0);
+                    result = writeColumn(column.name, *column.type, *permuted_column, offset_columns);
+                }
             }
             else
             {
-                /// We rearrange the columns that are not included in the primary key here; Then the result is released - to save RAM.
-                ColumnPtr permuted_column = column.column->permute(*permutation, 0);
-                writeColumn(column.name, *column.type, *permuted_column, offset_columns);
+                result = writeColumn(column.name, *column.type, *column.column, offset_columns);
             }
-        }
+            write_results[i] = result;
+        };
+
+        if (writing_thread_pool)
+            writing_thread_pool->scheduleOrThrowOnError(write_column_job);
```

Comment:
Also this threads will not be accounted as threads for query, to make them accounted you need to wrap `write_column_job` with:

```c
auto thread_group = CurrentThread::getGroup();
...

writing_thread_pool->scheduleOrThrowOnError([&write_column_job, thread_group]() {
    setThreadName("QueryPipelineEx");

    if (thread_group)
        CurrentThread::attachTo(thread_group);
    SCOPE_EXIT(
            if (thread_group)
                CurrentThread::detachQueryIfNotDetached();
    );

    write_column_job();
}
```

P.S. the code is completely untested